### PR TITLE
fix: don't vote on deposit requests in case of blocklist client errors

### PR DESCRIPTION
**Pull request by matteojug**
## Description

Closes: https://github.com/stacks-network/sbtc/issues/1130

## Changes

Add an immediate retry in case of blocklist client errors, and in case of two failures bubble up the error instead of voting against the deposit, so that we can try again at the next bitcoin block.

## Checklist:

- [ ] I have performed a self-review of my code
- [ ] My changes generate no new warnings
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: signer/src/request_decider.rs [hunk 1]
```diff
@@ -34,6 +36,9 @@ use crate::storage::DbWrite as _;
 use futures::StreamExt;
 use futures::TryStreamExt;
 
+/// Delay for the immediate retry in case of blocklist client errors
+const RETRY_DELAY: Duration = Duration::from_millis(100);
```
[the remark this answers is not in the document]
Comment (matteojug) [reply signer/src/request_decider.rs]: Yeah good point, added it.

File: signer/tests/integration/request_decider.rs [hunk 2]
```diff
@@ -398,3 +405,129 @@ async fn persist_received_deposit_decision_fetches_missing_deposit_requests() {
...
     testing::storage::drop_db(db).await;
 }
+
+/// Test `RequestDeciderEventLoop` behaviour in case of blocklist client
+/// failures. It should try to contact the blocklist client twice per bitcoin
+/// block, and in case of errors it should try again at the next block without
+/// voting.
+#[test_case::test_case(0, 0; "0 failures")]
+#[test_case::test_case(1, 0; "1 failure")]
+#[test_case::test_case(2, 1; "2 failures")]
```
Comment (djordon) [inline signer/tests/integration/request_decider.rs:415]: This is probably a bigger change, but can we add a test for the (2, 0) case where we end with zero rows in the database?

**Review by djordon — COMMENTED** [review]
Largely looks good.

We don't have a central place for all of our config defaults, but we should eventually converge on a place for them. `Settings::new` seems like it does the job but I don't have strong feeling about it.

File: signer/src/config/mod.rs [hunk 3]
```diff
@@ -212,8 +213,21 @@ pub struct BlocklistClientConfig {
...
     #[serde(deserialize_with = "url_deserializer_single")]
     pub endpoint: Url,
+
+    /// The delay, in milliseconds, for the second retry after a blocklist
+    /// client failure
+    #[serde(
+        default = "BlocklistClientConfig::retry_delay_default",
+        deserialize_with = "duration_milliseconds_deserializer"
+    )]
+    pub retry_delay: std::time::Duration,
```
[the remark this answers is not in the document]
Comment (matteojug) [reply signer/src/config/mod.rs:223]: The issue here is that if we set the default in `Settings::new` then we force the creation of `blocklist_client`, but without the url we get an error:
```
// doing something like:
//         cfg_builder = cfg_builder.set_default("blocklist_client.retry_delay", 100)?;

thread 'config::tests::default_config_toml_loads' panicked at signer/src/config/mod.rs:580:14:
Failed create settings from default config: missing field `endpoint`
```

File: signer/tests/integration/request_decider.rs (hunk 2 above)
Comment (matteojug) [reply signer/tests/integration/request_decider.rs:415]: We can add it, but IIUC this is already tested in the `#[test_case::test_case(2, 1; "2 failures")]` case, as for each failed iteration we check that we still have no votes (https://github.com/stacks-network/sbtc/pull/1270/files#diff-a8d19ef5983a5c5c6882f674d7b18dfd9326f0d7f25babfd863a986d000847a0R519). Are you referring to a different check?

Comment (djordon) [reply signer/tests/integration/request_decider.rs:415]: Ahh right! Yeah never mind.

File: signer/src/config/mod.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (djordon) [reply signer/src/config/mod.rs:223]: Ahhh I see. Okay then.

**Review by djordon — APPROVED** [review]
Looks good!
